cpp/include/cudf/column/column_device_view.cuh

```diff
@@ -558,6 +664,63 @@ class alignas(16) column_device_view : public detail::column_device_view_base {
                                                  detail::pair_rep_accessor<T, has_nulls>{*this}};
   }
 
+  /**
+   * @brief Return an optional iterator to the element following the last element of
+   * the column.
+   *
+   * Dereferencing the returned iterator returns a `thrust::optional<T>`.
+   *
+   * This function does not participate in overload resolution if
+   * `column_device_view::has_element_accessor<T>()` is false.
+   *
+   * @throws cudf::logic_error if the column is not nullable, and `DYNAMIC` mode used and
+   *         the user has stated nulls exist
+   * @throws cudf::logic_error if column datatype and Element type mismatch.
+   */
+  template <typename T, CUDF_ENABLE_IF(column_device_view::has_element_accessor<T>())>
+  auto optional_end(contains_nulls::DYNAMIC, bool has_nulls = true) const
```

Comment:
Hm, I'm wondering about if `has_nulls` should be defaulted or not. I tend to err on the side of being safer and requiring the caller to be explicit, but I could easily be convinced that defaulting it to true is the right thing.

Reply:
I think that being explicit with `DYNAMIC` makes sense. The default does have a foot-gun where you could state a columns has nulls when in fact it is non-nullable.

Maybe we want:
```cpp
auto optional_begin(contains_nulls::DYNAMIC, bool has_nulls) { ... }
auto optional_begin(contains_nulls::DYNAMIC) { 
  return optional_begin(contains_nulls::DYNAMIC, this->nullable());
}
```

Reply:
I have updated the changes to offer both signatures for `DYNAMIC`